cps/main.c

```diff
@@ -746,22 +818,45 @@ add_bgp_filters(struct gatekeeper_if *iface, uint16_t tcp_port_bgp,
 		/* Capture pkts for connections started by our BGP speaker. */
 		ret = ntuple_filter_add(iface->id, iface->ip4_addr.s_addr,
 			rte_cpu_to_be_16(tcp_port_bgp), UINT16_MAX, 0, 0,
-			IPPROTO_TCP, rx_queue, true);
+			IPPROTO_TCP, rx_queue, true, false);
 		if (ret < 0) {
-			RTE_LOG(ERR, GATEKEEPER,
-				"cps: could not add source BGP filter on %s iface\n",
-				iface->name);
-			return ret;
-		}
-		/* Capture pkts for connections remote BGP speakers started. */
-		ret = ntuple_filter_add(iface->id, iface->ip4_addr.s_addr,
-			0, 0, rte_cpu_to_be_16(tcp_port_bgp), UINT16_MAX,
-			IPPROTO_TCP, rx_queue, true);
-		if (ret < 0) {
-			RTE_LOG(ERR, GATEKEEPER,
-				"cps: could not add destination BGP filter on %s iface\n",
+			struct ipv4_acl_rule ipv4_rules[NUM_ACL_BGP_RULES];
+
+			RTE_LOG(NOTICE, PORT, "cps: filtering IPv4 BGP packets in software on the %s interface\n",
 				iface->name);
-			return ret;
+			iface->hw_filter_ntuple = false;
+
+			memset(&ipv4_rules, 0, sizeof(ipv4_rules));
+
+			/* Capture connections started by our BGP speaker. */
+			fill_bgp4_rule(&ipv4_rules[0], iface,
+				true, tcp_port_bgp);
+			/* Capture connections remote BGP speakers started. */
+			fill_bgp4_rule(&ipv4_rules[1], iface,
+				false, tcp_port_bgp);
+
+			ret = register_ipv4_acl(ipv4_rules, NUM_ACL_BGP_RULES,
+				submit_bgp, match_bgp4, iface);
+			if (ret < 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"cps: could not register BGP IPv4 ACL on %s iface\n",
+					iface->name);
+				return ret;
+			}
+		} else {
```

Comment:
I don't think we can do that here, because we only want to execute the code in the `else` if the ntuple filter is supported. If we get rid of the `else`, then after successfully adding an IPv4 ACL in the `if`, we would also try to add an ntuple filter.